core/src/shared_memory.rs

```diff
@@ -1,36 +1,45 @@
 use crate::syscalls;
+use core::cell::UnsafeCell;
 use core::ptr;
 
 #[must_use = "Shared memory risks being dropped too early. Drop it manually."]
-pub struct SharedMemory<'a> {
+pub struct SharedMemory<T> {
     driver_number: usize,
     allow_number: usize,
-    buffer_to_share: &'a mut [u8],
+    buffer_to_share: UnsafeCell<T>,
 }
 
-impl<'a> SharedMemory<'a> {
-    pub fn new(
-        driver_number: usize,
-        allow_number: usize,
-        buffer_to_share: &'a mut [u8],
-    ) -> SharedMemory<'a> {
+impl<T> SharedMemory<T>
+where
+    T: AsMut<[u8]>,
+{
+    pub fn new(driver_number: usize, allow_number: usize, buffer_to_share: T) -> SharedMemory<T> {
         SharedMemory {
             driver_number,
             allow_number,
-            buffer_to_share,
+            buffer_to_share: UnsafeCell::new(buffer_to_share),
         }
     }
 
-    pub fn read_bytes<T: AsMut<[u8]>>(&self, mut destination: T) {
-        safe_copy(self.buffer_to_share, destination.as_mut());
+    pub fn read_bytes<D: AsMut<[u8]>>(&self, mut destination: D) {
+        let buf = unsafe { (*self.buffer_to_share.get()).as_mut() };
+        safe_copy(buf, destination.as_mut());
     }
 
-    pub fn write_bytes<T: AsRef<[u8]>>(&mut self, source: T) {
-        safe_copy(source.as_ref(), self.buffer_to_share);
+    pub fn write_bytes<S: AsRef<[u8]>>(&mut self, source: S) {
+        let buf = unsafe { (*self.buffer_to_share.get()).as_mut() };
+        safe_copy(source.as_ref(), buf);
+    }
+
+    pub(crate) unsafe fn operate_on_mut_ptr<R: Sized, F: FnOnce(*mut u8) -> R>(
```

Comment:
- Couldn't this be transformed to a safe function where `F: FnOnce(&mut u8) ->R`?
- Shouldn't `R: Sized` be implicit?

Reply:
Right!